src/backend/distributed/replication/multi_logical_replication.c

```diff
@@ -291,33 +276,13 @@ LogicallyReplicateShards(List *shardList, char *sourceNodeName, int sourceNodePo
 	}
 	PG_CATCH();
 	{
+		/* We don't need to UnclaimConnections since we're already erroring out */
+
 		/*
-		 * Try our best not to leave any left-over subscription or publication.
-		 *
-		 * Although it is not very advisable to use code-paths that could throw
-		 * new errors, we prefer to do it here since we expect the cost of leaving
-		 * left-overs not be very low.
+		 * Drop temporary objects that were marked as CLEANUP_ON_FAILURE
+		 * or CLEANUP_ALWAYS.
 		 */
-
-		/* reconnect if the connection failed or is waiting for a command */
-		RecreateGroupedLogicalRepTargetsConnections(groupedLogicalRepTargetsHash,
-													superUser, databaseName);
-
-		DropSubscriptions(logicalRepTargetList);
-
-		/* reconnect if the connection failed or is waiting for a command */
-		if (PQstatus(sourceConnection->pgConn) != CONNECTION_OK ||
-			PQisBusy(sourceConnection->pgConn))
-		{
-			sourceConnection = GetNodeUserDatabaseConnection(connectionFlags,
-															 sourceNodeName,
-															 sourceNodePort, superUser,
-															 databaseName);
-		}
-		DropReplicationSlots(sourceConnection, logicalRepTargetList);
-		DropPublications(sourceConnection, publicationInfoHash);
-
-		/* We don't need to UnclaimConnections since we're already erroring out */
+		FinalizeOperationNeedingCleanupOnFailure("citus_[move/copy]_shard_placement");
```

Comment:
food for thought: Given that the maintenance daemon will clean these things up, why do we actively try to clean them up ourselves on failure? It actually causes some issues for users since the last release, because the cleanup itself might fail for some reason and that in turn masks the original error.

Reply:
If we don't clean them up in case of failure, wouldn't the next operation be blocked with an error like "subscription ... already exists"? 
But I agree with you. I think we can consider omitting the failure cleanup once we have unique object names for each operation.

Reply:
Sorry, just remembered that we use `CLEANUP_ON_FAILURE` only for shards, for now.
Still, I think it would be better to do this in a separate PR, as this change would require a lot of additional changes and make this PR more complex.

Reply:
CLEANUP_ALWAYS is also cleaned up by `FinalizeOperationNeedingCleanupOnFailure`. So lots of other things get cleaned up by it as well.


> If we don't clean them up in case of failure, wouldn't the next operation be blocked with an error like "subscription ... already exists"?

I think the main thing necessary to fix this is to change `DropOrphanedShardsInSeparateTransaction` to also run `CALL citus_cleanup_orphaned_resources()`

Reply:
> which I think is something that we should do anyway

Agree on that. But we would still be leaving shards to be deleted on worker nodes, in a failed shard split. And that could block the next shard split command executed right after the failed one. Shard move calls that failed leaves some replication objects behind as well, if we drop the failure cleanup. That would block the next shard move command as well.

I think we can apply your suggestion by adding some cleanup calls in regression tests. And maybe call the `DropOrphanedShardsInSeparateTransaction` in the code, at the beginning of each move/split operation.

But my main concern is that is it the way we want to cleanup things in case of failure? When we start giving unique names to replication objects; would we able to call the cleanup thing at the beginning of each operation? Maybe we can make it work by introducing some locks on the objects (to prevent operations from deleting each other's objects)

I have a slight preference on considering your suggestion apart from this pr, and applying it in a separate pr if we agree on doing it.